src/EmailNrGuardianVerifier.sol

```diff
@@ -0,0 +1,251 @@
+    function dkimRegistryAddr() public view returns (address) {
+        return address(dkimRegistry);
+    }
+
+    /**
+     * @dev Returns the address of the verifier contract.
+     *
+     * @return address The address of the verifier contract.
+     */
+    function verifierAddr() public view returns (address) {
+        return address(verifier);
+    }
+
+    /**
+     * @dev Verify the proof
+     * Recommended to be used when nullifier based check for replay protection are required & not handeled at higher level
+     * e.g. Email recovery functions ( handleAcceptance & handleRecovery )
+     *
+     * @notice Nullifier based check is handled here
+     * @notice Timestamp check of when the proof was generated is handled here
+     * @notice Reverts if the proof is invalid
+     *
+     * @param account Account to be recovered
+     * @param proof Proof data
+     * proof.data: EmailData
+     * proof.publicInputs: [publicKey, emailNullifier]
+     * proof.proof: zk-SNARK proof
+     *
+     * NOTE: Gas optimisation possible by only decoding the proof data once in this function rather in tryVerify as well
+     *
+     * @return isVerified if the proof is valid
+     */
+    function verifyProof(
+        address account,
+        ProofData memory proof
+    ) public returns (bool) {
+        EmailData memory emailData = abi.decode(proof.data, (EmailData));
+
+        bytes32 emailNullifier = proof.publicInputs[1];
+        require(
+            usedNullifiers[emailNullifier] == false,
+            EmailNullifierAlreadyUsed()
+        );
+
+        require(
+            timestampCheckEnabled == false ||
+                emailData.timestamp == 0 ||
+                emailData.timestamp > lastTimestamp,
+            InvalidTimestamp()
+        );
+
+        bool isVerified = tryVerifyProof(account, proof);
+
+        if (isVerified) {
+            usedNullifiers[emailNullifier] = true;
+            if (timestampCheckEnabled && emailData.timestamp != 0) {
+                lastTimestamp = emailData.timestamp;
+            }
+        }
+
+        return isVerified;
+    }
+
+    /**
+     * @dev Verify the proof
+     * Recommended to use when only proof verification is required
+     * View function to check if the proof is valid
+     *
+     * @notice Replay protection is assumed to be handled by the caller
+     * @notice Reverts if the proof is invalid
+     *
+     * @param account Account to be recovered
+     * @param proof Proof data
+     * proof.data: EmailData
+     * proof.publicInputs: [publicKey, emailNullifier]
+     * proof.proof: zk-SNARK proof
+     *
+     * @return isVerified if the proof is valid
+     */
+    function tryVerifyProof(
+        address account,
+        ProofData memory proof
+    ) public view returns (bool) {
+        // Parse the extra data
+        EmailData memory emailData = abi.decode(proof.data, (EmailData));
+
+        // TODO: How to handle dkimRegistry validation in case of noir circuits
```

Comment:
hmm good question. I'm not 100% certain but I would hope we can use it in the same way as for non-noir flow. Was there any reason it could not be implemented here?

Reply:
There was no information on how DKIM registry comes into the flow in case of noir circuits. Another catch here is the the prover generates a different publicKey everytime, so it might require to set public key again. I'm not too aware on the internals of it as well in zkemail.nr.

More info would help, but I doubt that can be done under the timeline.

Reply:
Understood. Could you mention this in the document for TODO `4. Collect blockers/issues with our stack into notes`

Reply:
And let's leave it as is